source/agora/network/NodeLocator.d

```diff
@@ -0,0 +1,268 @@
+        import std.utf : toUTF8;
+
+        switch (entry_data.type)
+        {
+            case MMDB_DATA_TYPE_UTF8_STRING :
+                return entry_data.utf8_string[0 .. entry_data.data_size].toUTF8();
+            case MMDB_DATA_TYPE_FLOAT :
+                return entry_data.float_value.to!string();
+            case MMDB_DATA_TYPE_DOUBLE :
+                return entry_data.double_value.to!string();
+            case MMDB_DATA_TYPE_UINT16 :
+                return entry_data.uint16.to!string();
+            case MMDB_DATA_TYPE_UINT32 :
+                return entry_data.uint32.to!string();
+            case MMDB_DATA_TYPE_INT32 :
+                return entry_data.int32.to!string();
+            case MMDB_DATA_TYPE_UINT64 :
+                return entry_data.uint64.to!string();
+            case MMDB_DATA_TYPE_BOOLEAN :
+                return entry_data.boolean.to!string();
+            default :
+                assert(0, "Datatype not implemented");
+        }
+    }
+}
+
+/// Node locator mock which is supposed to be used in network tests
+public class NodeLocatorMock : INodeLocator
+{
+    /// Start the node locator, returns true on succes, false otherwise
+    public override bool start () const @safe @nogc pure nothrow {return true;}
+
+    /// Stop the node locator
+    public override void stop () const @safe @nogc pure nothrow {}
+
+    /***************************************************************************
+
+        Extracts certain properties of the geographical location of the node.
+        The properties can be continent, country, ...
+        The exact properties that can be extracted and the query path that
+        identifies the properties are implementation dependent.
+
+        Params:
+            address = the network address of the node
+            paths = query parts to identify which properties needs to be extracted
+
+        Returns:
+            an array of extracted properties, if a particular property
+            cannot be extracted, then `MissingValue` is returned
+
+    ***************************************************************************/
+
+    public override string[] extractValues (Address address, string[] paths) const @safe pure nothrow
+    {
+        // Returns the same information regardless of the address
+        string[string] path_map =
+            [
+                "continent->names->en" : "Asia",
+                "country->names->en"   : "South Korea",
+                "city->names->en"     : "Seoul (Namdaemunno 5(o)-ga)",
+                "location->latitude"   : "1.111",
+                "location->longitude"  : "2.222",
+            ];
+
+        string[] extractedValues;
+        extractedValues.length = paths.length;
+
+        foreach (ind, path; paths)
+            if (auto it = path in path_map)
+                extractedValues[ind] = *it;
+            else
+                extractedValues[ind] = MissingValue;
+
+        return extractedValues;
+    }
```

Comment:
This looks like the wrong place to do dependency injection. The injection likely belongs to the place doing the data acquisition, not the place that transforms it.